src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -534,10 +538,30 @@ StoreQueryResult(PGconn *connection, TupleDesc tupleDescriptor,
 		resultStatus = PQresultStatus(result);
 		if ((resultStatus != PGRES_SINGLE_TUPLE) && (resultStatus != PGRES_TUPLES_OK))
 		{
-			WarnRemoteError(connection, result);
+			char *sqlStateString = PQresultErrorField(result, PG_DIAG_SQLSTATE);
+			int category = 0;
+			bool raiseError = false;
+
+			/*
+			 * If the error code is in constraint violation class, we want to
+			 * fail fast because we must get the same error from all shard
+			 * placements.
+			 */
+			category = ERRCODE_TO_CATEGORY(ERRCODE_INTEGRITY_CONSTRAINT_VIOLATION);
```

Comment:
I think that erroring out only for `ERRCODE_INTEGRITY_CONSTRAINT_VIOLATION` may lead to some cases where Citus marks placements inactive unnecessarily. For instance, if the error code is `ERRCODE_DIVISION_BY_ZERO` we got into this situation. 
Here are the steps:
```SQL

CREATE TABLE multiple_hash
(
	category text NOT NULL,
	data text NOT NULL 
);
SELECT master_create_distributed_table('multiple_hash', 'category', 'hash');
SELECT master_create_worker_shards('multiple_hash', 4, 2);

select * from pg_dist_shard_placement ;
 shardid | shardstate | shardlength | nodename  | nodeport 
---------+------------+-------------+-----------+----------
  102049 |          1 |           0 | localhost |     9700
  102049 |          1 |           0 | localhost |     9701
  102050 |          1 |           0 | localhost |     9701
  102050 |          1 |           0 | localhost |     9700
  102051 |          1 |           0 | localhost |     9700
  102051 |          1 |           0 | localhost |     9701
  102052 |          1 |           0 | localhost |     9701
  102052 |          1 |           0 | localhost |     9700
(8 rows)

-- insert a single value
INSERT INTO multiple_hash VALUES ('1', '1');

-- do a returning where we end-up with division by zero error
update multiple_hash SET data = (5 / (category::int - 1)) where category = '1' RETURNING *;
WARNING:  division by zero
CONTEXT:  while executing command on localhost:9701
WARNING:  division by zero
CONTEXT:  while executing command on localhost:9700
 category | data 
----------+------
(0 rows)

-- select again placement table
select * from pg_dist_shard_placement ;
 shardid | shardstate | shardlength | nodename  | nodeport 
---------+------------+-------------+-----------+----------
  102045 |          1 |           0 | localhost |     9700
  102045 |          1 |           0 | localhost |     9701
  102046 |          1 |           0 | localhost |     9701
  102047 |          1 |           0 | localhost |     9700
  102047 |          1 |           0 | localhost |     9701
  102048 |          1 |           0 | localhost |     9701
  102048 |          1 |           0 | localhost |     9700
  102046 |          3 |           0 | localhost |     9700
(8 rows)

```
Not that this doesn't happen without returning statement. I think the reason is that `StoreQueryResult()` returns true and  `ConsumeQueryResult()` returns false in the above case.